### PR TITLE
Added lstat fallback and Haiku WASI support

### PR DESCRIPTION
**Pull request by SamuraiCrow**
Some POSIX filesystems don't support d_type under Linux and some operating systems, like Haiku, don't even supply d_type in their implementation of dirent at all.  It is required to do lstat whenever DT_UNKNOWN is supplied by the filesystem or when there is no d_type shortcut.  Also, the test had a few shortcomings on Manjaro Linux.

### CODE REVIEW

**Comment by SamuraiCrow** [comment]
Currently compiles with Clang on both Haiku and Manjaro Linux.  GCC 11 and 12 don't compile yet but appears to be a compiler bug in those versions of GCC.

File: wasi/wasi.c [hunk 1]
```diff
@@ -1042,6 +1044,7 @@ wasiFDReaddir(
     U32 bufferUsed = 0;
     U64 next = 0;
     U64 inode = 0;
+	struct stat entry_stat;
```
Comment (korli) [inline wasi/wasi.c:1047]: Looks like a tab vs spaces.

File: wasi/wasi.c [hunk 2]
```diff
@@ -1104,8 +1107,33 @@ wasiFDReaddir(
         inode = entry->d_ino;
         name = entry->d_name;
         nameLength = strlen(name);
-        fileType = wasiFileTypeFromDirentFileType(entry->d_type);
+#ifndef __HAIKU__
+        if (entry->d_type != DT_UNKNOWN) {
+			fileType = wasiFileTypeFromDirentFileType(entry->d_type);
```
Comment (korli) [inline wasi/wasi.c:1112]: here too tab vs spaces.

File: wasi/wasi.c (hunk 1 above)
Comment (SamuraiCrow) [reply wasi/wasi.c]: Fixed.

File: wasi/wasi.c (hunk 2 above)
Comment (SamuraiCrow) [reply wasi/wasi.c]: Also Fixed.

**Comment by turbolent** [comment]
Thank you for the PR @SamuraiCrow! 

I didn't realize `d_type` is non-POSIX, should have read the man page more thoroughly. 

Instead of making this Haiku specific, maybe we can aim for something more general? LLVM just checks for `DTTOIF`: https://reviews.llvm.org/D64940. In addition using it for checking the existence of `d_type`, we should also be able to then use it to translate the type to the mode and use `wasiFileTypeFromMode`.

File: wasi/Makefile [hunk 3]
```diff
@@ -1,4 +1,5 @@
-CFLAGS += -std=c89 -Wunused-result -Wall -Wpedantic -Wno-long-long -Wno-unused-function
+# TODO Find out why limits.h doesn't define PATH_MAX on Manjaro Linux
```
Comment (turbolent) [inline wasi/Makefile:1]: As far as I understand, `PATH_MAX` is not defined for `-std=c89`

File: wasi/test.c [hunk 4]
```diff
@@ -39,10 +39,11 @@ testResolvePath(
     fprintf(stderr, "OK: resolvePath(%s, %s) == %s\n", directory, path, expected);
 }
 
+wasmMemory wasiMemory;
+
 int
 main(int argc, char* argv[]) {
-    wasmMemory m;
```
Comment (turbolent) [inline wasi/test.c:44]: Any particular reason for making this global?

File: wasi/wasi.c [hunk 5]
```diff
@@ -1104,8 +1107,33 @@ wasiFDReaddir(
...
+        if (entry->d_type != DT_UNKNOWN) {
+            fileType = wasiFileTypeFromDirentFileType(entry->d_type);
+        } else {
+#endif
+            if (lstat(descriptor.path, &entry_stat)) {
+                WASI_TRACE(("fd_readdir: lstat failed: %s", strerror(errno)));
+                return wasiErrno();
+            }
 
+            if (S_ISREG(entry_stat.st_mode)) {
```
Comment (turbolent) [inline wasi/wasi.c:1120]: You can use the existing function `wasiFileTypeFromMode` here

File: wasi/Makefile (hunk 3 above)
Comment (SamuraiCrow) [reply wasi/Makefile:1]: Ah yes.  That would explain it.

File: wasi/wasi.c (hunk 5 above)
Comment (SamuraiCrow) [reply wasi/wasi.c:1120]: Good idea.  I'll make that change right away.

**Comment by SamuraiCrow** [comment]
I've made the DTTOIF macro support present but guarded by `#if defined(DTTOIF)` like LLVM has done.

**Comment by SamuraiCrow** [comment]
I've also moved `wasiFileTypeFromMode()` up to where `wasiFileTypeFromDirent()` was so there's no need for a function prototype.

File: wasi/wasi.c [hunk 6]
```diff
@@ -1104,8 +1108,17 @@ wasiFDReaddir(
...
+        fileType = wasiFileTypeFromMode(DTTOIF(entry->d_type));
+#else
+        if (lstat(descriptor.path, &entry_stat)) {
+            WASI_TRACE(("fd_readdir: lstat failed: %s", strerror(errno)));
+            return wasiErrno();
+        }
 
+        fileType = wasiFileTypeFromMode(entry_stat.st_mode);
+
+#endif
```
Comment (turbolent) [inline wasi/wasi.c:1121]: Nice, this handles the case where d_type is not available in a more portable way 👍 

We should also handle the case where d_type is available, but is not supported by the file system, maybe like so: 

```suggestion
        // Some operating systems do not provide d_type (Linux, macOS, and some BSDs do).
        // When available, d_type is not supported by all filesystem types.
        // Fall back by using lstat.

#if defined (DTTOIF)
        fileType = wasiFileTypeFromMode(DTTOIF(entry->d_type));
#else
        fileType = wasiFileTypeUnknown;
#endif

        if (fileType == wasiFileTypeUnknown) {
            struct stat entryStat;

            if (lstat(descriptor.path, &entryStat)) {
                WASI_TRACE(("fd_readdir: lstat failed: %s", strerror(errno)));
                return wasiErrno();
            }

            fileType = wasiFileTypeFromMode(entryStat.st_mode);
        }
```

File: wasi/wasi.c [hunk 7]
```diff
@@ -1042,6 +1043,9 @@ wasiFDReaddir(
     U32 bufferUsed = 0;
     U64 next = 0;
     U64 inode = 0;
+#if !defined(DTTOIF)
+    struct stat entry_stat;
+#endif
```
Comment (turbolent) [inline wasi/wasi.c:1048]: This can be defined in case it is needed, see below:
```suggestion
```

**Review by turbolent — COMMENTED** [review]
Looks good!

File: wasi/wasi.c (hunk 6 above)
Comment (SamuraiCrow) [reply wasi/wasi.c:1121]: I see that `fileType` is defined at the beginning of the function as wasiFileTypeUnknown so no need to redefine it if the macro is not available.  Also, it is not redefined until this point anyway.  Futhermore, the double slash comment type was added in C99 and C++, not present in C89.  The rest looks ok.  I'll commit this with the changes I just mentioned.

Comment (turbolent) [reply wasi/wasi.c]: It is defined at the beginning of the function, but re-assigned in the while loop, so should be reset in each iteration (call of `readdir`).

Good catch on the C++ comment syntax.

Please use the syntax as suggested (spacing between operators, camelCase for variable name

File: wasi/Makefile (hunk 3 above)
Comment (turbolent) [reply wasi/Makefile:1]: Can you please remove this TODO? I'll try to address this in a follow-up PR

File: wasi/test.c (hunk 4 above)
Comment (turbolent) [reply wasi/test.c:44]: Ah, I see. `wasiMemory` is expected to be a function that the WASI implementation can call. 
Thanks for noticing this! 

See the suggestion above

File: wasi/test.c [hunk 8]
```diff
@@ -39,10 +39,11 @@ testResolvePath(
     fprintf(stderr, "OK: resolvePath(%s, %s) == %s\n", directory, path, expected);
 }
 
+wasmMemory wasiMemory;
```
Comment (turbolent) [inline wasi/test.c:42]:
```suggestion
/* unused, but expected by the WASI implementation */
wasmMemory* wasiMemory(void* instance) {
	return NULL
}
```

**Review by turbolent — COMMENTED** [review]
Feel free to open as ready for review

File: wasi/test.c (hunk 8 above)
Comment (SamuraiCrow) [reply wasi/test.c:42]: Fixed!

File: wasi/Makefile [hunk 9]
```diff
@@ -1,4 +1,4 @@
-CFLAGS += -std=c89 -Wunused-result -Wall -Wpedantic -Wno-long-long -Wno-unused-function
+CFLAGS += -std=c89 -Wunused-result -Wall -Wno-long-long -Wno-unused-function -DPATH_MAX=65536
```
Comment (turbolent) [inline wasi/Makefile:1]:
```suggestion
CFLAGS += -std=c89 -Wunused-result -Wall -Wpedantic -Wno-long-long -Wno-unused-function
```

**Review by turbolent — APPROVED** [review]
Thank you!

**Pull request merged**